File: cli/src/commands/session/transcript.ts
import { Command } from "commander";
import { daemonGet } from "../../lib/daemon-client.js";
import { preflight } from "../../lib/preflight.js";
import { die } from "../../lib/output.js";

interface NormalizedEvent {
  id: string;
  kind: string;
  role?: string;
  text?: string;
  toolName?: string;
  toolResult?: { content?: string; isError?: boolean };
}

interface TranscriptResponse {
  events: NormalizedEvent[];
}

export function registerSessionTranscript(session: Command): void {
  session
    .command("transcript <id>")
    .description("Print a JSON session's normalized transcript")
    .option("--json", "Output raw NDJSON events")
    .action(async (id: string, opts: { json?: boolean }) => {
      await preflight();

      const result = await daemonGet<TranscriptResponse>(`/sessions/${id}/transcript`);
      if (!result.ok) {
        die(result.error, result.status === 404 ? 2 : 1);
      }

      const events = result.data.events ?? [];
      if (opts.json) {
        for (const ev of events) {
          console.log(JSON.stringify(ev));
        }
        process.exit(0);
      }

      if (events.length === 0) {
        console.log("(no transcript yet)");
        return;
      }
      for (const ev of events) {
        const who = ev.role ? `${ev.kind}/${ev.role}` : ev.kind;
        const body =
          ev.text ??
          ev.toolName ??
          ev.toolResult?.content ??
          "";
        const oneLine = body.replace(/\s+/g, " ").slice(0, 200);
        console.log(`[${who}] ${oneLine}`);
      }
    });
}

File: rust/vst-cli/src/commands/session/transcript.rs
use std::process;
use vst_types::rest::sessions::AllEvents;

use crate::client::{daemon_get, DaemonResult};
use crate::preflight::preflight;

#[derive(Clone, Debug, Default, PartialEq)]
pub struct SessionTranscriptOptions {
    pub id: String,
    pub json: bool,
}

pub fn parse_session_transcript_options(
    args: &[String],
) -> Result<SessionTranscriptOptions, String> {
    let mut opts = SessionTranscriptOptions::default();
    let mut positional = Vec::new();

    for arg in args {
        match arg.as_str() {
            "--json" => {
                opts.json = true;
            }
            other if other.starts_with('-') => {
                return Err(format!("Unknown option: {other}"));
            }
            other => {
                positional.push(other.to_string());
            }
        }
    }

    if positional.is_empty() {
        return Err("Session ID is required".to_string());
    }
    opts.id = positional[0].clone();

    Ok(opts)
}

pub async fn run_session_transcript(opts: SessionTranscriptOptions) -> Result<(), (String, i32)> {
    preflight().await;

    let result = daemon_get::<AllEvents>(&format!("/sessions/{}/transcript", opts.id))
        .await
        .map_err(|e| (e.to_string(), 1))?;

    match result {
        DaemonResult::Ok { data, .. } => {
            let events = data.events;
            if opts.json {
                for ev in events {
                    if let Ok(line) = serde_json::to_string(&ev) {
                        println!("{line}");
                    }
                }
                process::exit(0);
            }

            if events.is_empty() {
                println!("(no transcript yet)");
                return Ok(());
            }

            for ev in events {
                let kind_str = serde_json::to_value(ev.kind)
                    .ok()
                    .and_then(|v| v.as_str().map(ToString::to_string))
                    .unwrap_or_default();
                let role_str = ev
                    .role
                    .and_then(|r| serde_json::to_value(r).ok())
                    .and_then(|v| v.as_str().map(ToString::to_string));

                let who = if let Some(r) = role_str {
                    format!("{kind_str}/{r}")
                } else {
                    kind_str
                };

                let body = ev
                    .text
                    .or(ev.tool_name)
                    .or_else(|| ev.tool_result.and_then(|tr| tr.content))
                    .unwrap_or_default();

                let one_line = body.split_whitespace().collect::<Vec<_>>().join(" ");
                let truncated: String = one_line.chars().take(200).collect();
                println!("[{who}] {truncated}");
            }

            Ok(())
        }
        DaemonResult::Err { status, error, .. } => {
            let code = if status == 404 { 2 } else { 1 };
            Err((error, code))
        }
    }
}
